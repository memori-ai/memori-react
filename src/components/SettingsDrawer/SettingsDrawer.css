.memori-settings-drawer .memori-settings-drawer--field {
    display: flex;
    flex-direction: column;
    margin: 0.5rem 0;
}

.memori-settings-drawer .memori-settings-drawer--field .memori-select--value {
    color: #000;
}

.memori-settings-drawer--controlsposition-radio,
.memori-settings-drawer--avatarType-radio,
.memori-settings-drawer--microphoneMode-radio {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    margin: 0.5rem 0;
    gap: 0.2rem;
}
