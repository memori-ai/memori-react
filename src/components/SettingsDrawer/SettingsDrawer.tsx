import Drawer from '../ui/Drawer';
import { useTranslation } from 'react-i18next';
import Checkbox from '../ui/Checkbox';
import Select from '../ui/Select';
import { setLocalConfig } from '../../helpers/configuration';
import { RadioGroup, Switch } from '@headlessui/react';
import Button from '../ui/Button';
import { Props as WidgetProps } from '../MemoriWidget/MemoriWidget';
import { useState } from 'react';
import Slider from '../ui/Slider';
export interface Props {
  open: boolean;
  layout?: WidgetProps['layout'];
  onClose: () => void;
  microphoneMode?: 'HOLD_TO_TALK' | 'CONTINUOUS';
  continuousSpeechTimeout?: number;
  setMicrophoneMode: (value: 'HOLD_TO_TALK' | 'CONTINUOUS') => void;
  setContinuousSpeechTimeout: (value: number) => void;
  controlsPosition?: 'center' | 'bottom';
  setControlsPosition: (value: 'center' | 'bottom') => void;
  hideEmissions?: boolean;
  setHideEmissions: (value: boolean) => void;
  additionalSettings?: WidgetProps['additionalSettings'];
  avatarType?: 'blob' | 'avatar3d' | null;
  setAvatarType: (value: 'blob' | 'avatar3d' | null) => void;
  enablePositionControls?: boolean;
  setEnablePositionControls: (value: boolean) => void;
  isAvatar3d?: boolean;
}

const silenceSeconds = [2, 3, 5, 10, 15, 20, 30, 60];

const SettingsDrawer = ({
  open,
  layout = 'DEFAULT',
  onClose,
  microphoneMode = 'HOLD_TO_TALK',
  continuousSpeechTimeout,
  setMicrophoneMode,
  setContinuousSpeechTimeout,
  controlsPosition,
  setControlsPosition,
  hideEmissions,
  setHideEmissions,
  additionalSettings,
  avatarType,
  setAvatarType,
  enablePositionControls,
  setEnablePositionControls,
  isAvatar3d,
}: Props) => {
  const { t } = useTranslation();

  return (
    <Drawer
      className="memori-settings-drawer"
      open={open}
      onClose={onClose}
      title={t('widget.settings') || 'Settings'}
      description={t('write_and_speak.settingsHeaderLabel')}
    >
      <div className="memori-settings-drawer--field controls">
        <label htmlFor="#microphoneMode">
          {t('write_and_speak.microphoneMode') || 'Microphone mode'}:
        </label>
        <RadioGroup
          id="microphoneMode"
          name="microphoneMode"
          value={microphoneMode}
          defaultValue={microphoneMode}
          className="memori-settings-drawer--microphoneMode-radio"
          onChange={(value: any) => {
            let micMode =
              value === 'CONTINUOUS' ? 'CONTINUOUS' : 'HOLD_TO_TALK';

            setMicrophoneMode(micMode as 'CONTINUOUS' | 'HOLD_TO_TALK');
            setLocalConfig('microphoneMode', micMode);
          }}
        >
          <RadioGroup.Option
            value="HOLD_TO_TALK"
            className="memori-settings-drawer--microphoneMode-radio-button"
          >
            {({ checked }) => (
              <Button primary={checked} outlined={!checked}>
                {t('write_and_speak.holdToSpeak') || 'Hold to speak'}
              </Button>
            )}
          </RadioGroup.Option>
          <RadioGroup.Option
            value="CONTINUOUS"
            className="memori-settings-drawer--microphoneMode-radio-button"
          >
            {({ checked }) => (
              <Button primary={checked} outlined={!checked}>
                {t('write_and_speak.continuousSpeechLabel') ||
                  'Continuous speech'}
              </Button>
            )}
          </RadioGroup.Option>
        </RadioGroup>
      </div>

      {microphoneMode === 'CONTINUOUS' && (
        <div className="memori-settings-drawer--field">
          <Select
            label={t('write_and_speak.secondsLabel') || 'Seconds'}
            placeholder={t('write_and_speak.secondsLabel') || 'Seconds'}
            options={silenceSeconds.map(s => ({ value: s, label: s }))}
            value={continuousSpeechTimeout}
            onChange={value => {
              setContinuousSpeechTimeout(value);
              setLocalConfig('continuousSpeechTimeout', value);
            }}
          />
        </div>
      )}

      {layout === 'TOTEM' && (
        <>
          <div className="memori-settings-drawer--field controls">
            <label htmlFor="#controlsPosition">
              {t('write_and_speak.controlsPosition') || 'Controls'}:
            </label>
            <RadioGroup
              id="controlsPosition"
              name="controlsPosition"
              value={controlsPosition}
              defaultValue={controlsPosition}
              className="memori-settings-drawer--controlsposition-radio"
              onChange={(value: any) => {
                setControlsPosition(value);
                setLocalConfig('controlsPosition', value);
              }}
            >
              <RadioGroup.Option
                value="center"
                className="memori-settings-drawer--controlsposition-radio-button"
              >
                {({ checked }) => (
                  <Button primary={checked} outlined={!checked}>
                    {t('center') || 'Center'}
                  </Button>
                )}
              </RadioGroup.Option>
              <RadioGroup.Option
                value="bottom"
                className="memori-settings-drawer--controlsposition-radio-button"
              >
                {({ checked }) => (
                  <Button primary={checked} outlined={!checked}>
                    {t('bottom') || 'Bottom'}
                  </Button>
                )}
              </RadioGroup.Option>
            </RadioGroup>
          </div>

          {isAvatar3d && (
            <>
              <div className="memori-settings-drawer--field controls">

                <label
                  htmlFor="avatarType"
                  className="memori-settings-drawer-label"
                >
                  {t('write_and_speak.avatarType') || 'Avatar type'}:
                </label>
                <RadioGroup
                  id="avatarType"
                  name="avatarType"
                  value={avatarType}
                  defaultValue={avatarType}
                  className="memori-settings-drawer--avatarType-radio"
                  onChange={(value: any) => {
                    setAvatarType && setAvatarType(value);
                    setLocalConfig('avatarType', value);
                  }}
                >
                  <RadioGroup.Option
                    value="blob"
                    className="memori-settings-drawer--avatarType-radio-button"
                  >
                    {({ checked }) => (
                      <Button primary={checked} outlined={!checked}>
                        {t('write_and_speak.blob') || 'Blob'}
                      </Button>
                    )}
                  </RadioGroup.Option>
                  <RadioGroup.Option
                    value="avatar3d"
                    className="memori-settings-drawer--avatarType-radio-button"
                  >
                    {({ checked }) => (
                      <Button primary={checked} outlined={!checked}>
                        {t('write_and_speak.avatar3d') || 'Avatar 3D'}
                      </Button>
                    )}
                  </RadioGroup.Option>
                </RadioGroup>
              </div>

              <div className="memori-settings-drawer--field">
                <Checkbox
                  label={
                    t('write_and_speak.enablePositionControls') ||
                    'Enable position controls'
                  }
                  name="enablePositionControls"
                  checked={enablePositionControls}
                  onChange={e => {
                    setEnablePositionControls(e.target.checked);
                  }}
                />
              </div>
            </>
          )}

          <div className="memori-settings-drawer--field">
            <Checkbox
              label={
                t('write_and_speak.hideEmissionsLabel') || 'Hide emissions'
              }
              name="hideControls"
              checked={hideEmissions}
              onChange={e => {
                setHideEmissions(e.target.checked);
                setLocalConfig('hideEmissions', e.target.checked);
              }}
            />
          </div>
        </>
      )}

      {additionalSettings && (
        <div className="memori-settings-drawer--field controls">
          {additionalSettings}
        </div>
      )}
    </Drawer>
  );
};

export default SettingsDrawer;
