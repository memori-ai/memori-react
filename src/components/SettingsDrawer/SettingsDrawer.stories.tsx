import React from 'react';
import { Meta, Story } from '@storybook/react';
import I18nWrapper from '../../I18nWrapper';
import SettingsDrawer, { Props } from './SettingsDrawer';
import './SettingsDrawer.css';

const meta: Meta = {
  title: 'Widget/SettingsDrawer',
  component: SettingsDrawer,
  argTypes: {
    open: {
      control: {
        type: 'boolean',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<Props> = args => {
  const [microphoneMode, setMicrophoneMode] = React.useState<
    'HOLD_TO_TALK' | 'CONTINUOUS'
  >('HOLD_TO_TALK');
  const [controlsPosition, setControlsPosition] = React.useState<
    'bottom' | 'center'
  >('bottom');

  return (
    <I18nWrapper>
      <SettingsDrawer
        {...args}
        microphoneMode={microphoneMode}
        setMicrophoneMode={setMicrophoneMode}
        controlsPosition={controlsPosition}
        setControlsPosition={setControlsPosition}
      />
    </I18nWrapper>
  );
};

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  open: true,
  onClose: () => {},
  continuousSpeech: false,
  continuousSpeechTimeout: 2,
  setContinuousSpeech: () => {},
  setContinuousSpeechTimeout: () => {},
};

export const WithinTotemLayout = Template.bind({});
WithinTotemLayout.args = {
  layout: 'TOTEM',
  open: true,
  onClose: () => {},
  continuousSpeech: false,
  continuousSpeechTimeout: 2,
  setContinuousSpeech: () => {},
  setContinuousSpeechTimeout: () => {},
};

const AdditionalSettings = () => {
  const [hideEmissions, setHideEmissions] = React.useState(false);

  return (
    <div>
      <label htmlFor="#hideEmissions">Hide emissions:</label>
      <input
        id="hideEmissions"
        name="hideEmissions"
        type="checkbox"
        checked={hideEmissions}
        onChange={e => setHideEmissions(e.target.checked)}
      />
    </div>
  );
};

export const WithAdditionalSettings = Template.bind({});
WithAdditionalSettings.args = {
  open: true,
  onClose: () => {},
  continuousSpeech: false,
  continuousSpeechTimeout: 2,
  setContinuousSpeech: () => {},
  setContinuousSpeechTimeout: () => {},
  additionalSettings: <AdditionalSettings />,
};
