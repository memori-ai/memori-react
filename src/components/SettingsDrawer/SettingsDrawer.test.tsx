import { render } from '@testing-library/react';
import SettingsDrawer from './SettingsDrawer';
import React from 'react';

beforeEach(() => {
  // @ts-ignore
  window.IntersectionObserver = jest.fn(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  }));
});

it('renders SettingsDrawer unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      open={false}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer open unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      open={true}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer open with continuous speech enabled unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      open={true}
      onClose={jest.fn()}
      microphoneMode="CONTINUOUS"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer open with non-default continuous speech timeout unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      open={true}
      onClose={jest.fn()}
      microphoneMode="CONTINUOUS"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={10}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer for totem layout open unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      layout="TOTEM"
      open={true}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer for totem layout open with controls at center unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      layout="TOTEM"
      open={true}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="center"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders SettingsDrawer for totem layout with continuous speech and hide emissions unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      layout="TOTEM"
      open={true}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={true}
      setHideEmissions={jest.fn()}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

const AdditionalSettings = () => {
  const [hideEmissions, setHideEmissions] = React.useState(false);

  return (
    <div>
      <label htmlFor="#hideEmissions">Hide emissions:</label>
      <input
        id="hideEmissions"
        name="hideEmissions"
        type="checkbox"
        checked={hideEmissions}
        onChange={e => setHideEmissions(e.target.checked)}
      />
    </div>
  );
};

it('renders SettingsDrawer with additional custom settings unchanged', () => {
  const { container } = render(
    <SettingsDrawer
      open={true}
      onClose={jest.fn()}
      microphoneMode="HOLD_TO_TALK"
      setMicrophoneMode={jest.fn()}
      continuousSpeechTimeout={2}
      setContinuousSpeechTimeout={jest.fn()}
      controlsPosition="bottom"
      setControlsPosition={jest.fn()}
      hideEmissions={false}
      setHideEmissions={jest.fn()}
      additionalSettings={<AdditionalSettings />}
      avatarType="avatar3d"
      setAvatarType={jest.fn()}
      setEnablePositionControls={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});
