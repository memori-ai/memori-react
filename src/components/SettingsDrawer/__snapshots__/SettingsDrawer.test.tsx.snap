// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SettingsDrawer for totem layout open unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer for totem layout open with controls at center unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer for totem layout with continuous speech and hide emissions unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer open unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer open with continuous speech enabled unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer open with non-default continuous speech timeout unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders SettingsDrawer unchanged 1`] = `<div />`;

exports[`renders SettingsDrawer with additional custom settings unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
